pkg/serial/handshake.go

```diff
@@ -60,58 +60,57 @@ func HandshakeClient(ctx context.Context, conn net.Conn, debug bool) error {
 	if tracing {
 		defer trace.End(trace.Begin(""))
 	}
-	syn := make([]byte, 2)
-	synack := make([]byte, 2)
-	ack := make([]byte, 2)
-
+	synAck := make([]byte, 3)
 	buf := make([]byte, 255)
-
-	syn[0] = flagSyn
-	synack[0] = flagAck
-	ack[0] = flagAck
+	syn := []byte{flagSyn, 0}
+	rand.Read(syn[1:])
 
 	// set the read deadline for timeout
 	// this has no effect on windows as the deadline is set at port open time
 	deadline, ok := ctx.Deadline()
 	if ok {
+		log.Debugf("Setting deadline to receive data from server: %s", deadline)
 		conn.SetReadDeadline(deadline)
 	}
 
-	rand.Read(syn[1:])
-
+	log.Debugf("HandshakeClient: Writing SYN: %#v", syn)
 	conn.Write(syn)
 
-	if n, err := io.ReadFull(conn, buf[:3]); n != 3 || err != nil {
-
-		if n == 0 && err != nil {
-			return err
-		}
-
-		msg := fmt.Sprintf("HandshakeClient: failed to read expected SYN-ACK: n=%d, err=%s buf=[%#x]", n, err, buf[:n])
+	if n, err := io.ReadFull(conn, synAck); err != nil {
+		log.Errorf("HandshakeClient: failed to read expected SYN-ACK: n=%d, err=%s buf=[%#x]",
+			n, err, buf[:n])
 		if err != nil {
-			log.Error(msg)
-		} else {
-			log.Debug(msg)
+			return err
 		}
-		return err
 	}
 
-	synack[1] = syn[1] + 1
-	if bytes.Compare(synack, buf[:2]) != 0 {
-		msg := fmt.Sprintf("HandshakeClient: did not receive synack: %#x != %#x", synack, buf[:2])
-		log.Debugf(msg)
+	log.Debugf("Received SYNACK: %#v", synAck)
+	if synAck[0] != flagAck || synAck[1] != syn[1]+1 {
 		conn.Write([]byte{flagNak})
-		return errors.New(msg)
+		return fmt.Errorf("Unexpected sequence received for SYNACK: %#v", synAck)
 	}
 
-	log.Debugf("HandshakeClient: received synack: %#x == %#x\n", synack, buf[:2])
+	ack := []byte{flagAck, synAck[2] + 1}
 	log.Debug("client: writing ack")
-	ack[1] = buf[2] + 1
 	if debug {
 		ack[0] = flagDebugAck
 	}
 	conn.Write(ack)
 
+	// Ensure both sides are in sync. Both sides will send the same message
```

Comment:
Why do this at all? I cannot see that it adds any value.
If we've somehow stumbled into a correct handshake sequence that's still misaligned then the SSH protocol handshake will fail (which is essentially sending the same known string on both side in exactly the same way this is) and we'll retry from the top, the same way we do here.


The lossiness check below was added to assert that we weren't losing data during debugging but this is essentially a second handshake